lib/message-handler.js

```diff
@@ -177,539 +183,540 @@ class MessageHandler {
+                                    }
+
+                                    auditPromises.push(
+                                        this.auditHandler.store(auditData._id, raw, {
+                                            date: messageData.idate,
+                                            msgid: messageData.msgid,
+                                            header: messageData.mimeTree && messageData.mimeTree.parsedHeader,
+                                            ha: messageData.ha,
+                                            mailbox: mailboxData._id,
+                                            mailboxPath: mailboxData.path,
+                                            info: Object.assign({ queueId: messageData.outbound }, messageData.meta)
+                                        })
+                                    );
                                 }
 
-                                newPrepared.id = prepared.id; // retain original
+                                await Promise.all(auditPromises);
+                            };
 
-                                options.prepared = newPrepared; // new prepared in options just in case
-                                prepared = newPrepared; // overwrite top-level original prepared
-                                options.maildata = this.indexer.getMaildata(newPrepared.mimeTree); // get new maildata of encrypted message
-                                addMessage();
+                            // can safely cleanup, no err given. Returns pending promise, which is fine
+                            const cleanupRes = cleanup(null, true, {
+                                uidValidity,
+                                uid,
+                                id: messageData._id.toString(),
+                                mailbox: mailboxData._id.toString(),
+                                mailboxPath: mailboxData.path,
+                                size,
+                                status: 'new'
                             });
-                        });
-                    } else {
-                        // not already encrypted and no need to
-                        this.prepareMessage(options, (err, newPrepared) => {
-                            if (err) {
-                                return callback(err);
-                            }
 
-                            prepared = newPrepared;
-                            addMessage();
-                        });
-                    }
-                } else {
-                    // message already encrypted
-                    this.prepareMessage(options, (err, newPrepared) => {
-                        if (err) {
-                            return callback(err);
+                            // do not use more suitable finally as it is not supported in Node v8
+                            processAudits().then(resolve(cleanupRes)).catch(resolve(cleanupRes));
```

Comment:
You are triggering `resolve()` here before `processAudits` is finished.

Use this snippet to test function ordering:

```js
function main(){
    return new Promise((resolve, reject)=>{

        let someAction = async () =>{
            await new Promise(r=>setTimeout(r, 100))
            console.log('some action');
        };
        someAction().then(resolve(1)).catch(reject(2))
    });
}
main().then(console.log).catch(console.error);
```

This outputs
```
1
some action
```